projects/driver_controls/src/magnetic_brake_event_generator.c

```diff
@@ -0,0 +1,138 @@
+#include <limits.h>
+#include <stdint.h>
+#include <stdio.h>
+
+#include <stdbool.h>
+#include "adc.h"
+#include "ads1015.h"
+#include "ads1015_def.h"
+#include "delay.h"
+#include "event_arbiter.h"
+#include "event_queue.h"
+#include "gpio_it.h"
+#include "i2c.h"
+#include "input_event.h"
+#include "interrupt.h"
+#include "log.h"
+#include "magnetic_brake_event_generator.h"
+#include "soft_timer.h"
+#include "status.h"
+#include "unity.h"
+
+int16_t percentage_converter(MagneticCalibrationData *data, MagneticBrakeSettings *brake_settings) {
+  int16_t percentage;
+
+  if (brake_settings->zero_value > brake_settings->hundred_value) {
+    percentage =
+        ((brake_settings->min_allowed_range * (data->reading - brake_settings->hundred_value)) /
+         (brake_settings->hundred_value - brake_settings->zero_value)) +
+        brake_settings->max_allowed_range;
+  } else {
+    percentage =
+        (brake_settings->max_allowed_range * (data->reading - brake_settings->zero_value)) /
+        (brake_settings->hundred_value - brake_settings->zero_value);
+  }
+
+  if (percentage < brake_settings->min_allowed_range) {
+    percentage = brake_settings->min_allowed_range;
+  } else if (percentage > brake_settings->max_allowed_range) {
+    percentage = brake_settings->max_allowed_range;
+  }
+
+  // add code to see if the percentage is greater than max allowed value
+  if (percentage > brake_settings->percentage_threshold) {
+    uint16_t percentage_data = (uint16_t)percentage;
```

Comment:
You should validate this cast is safe. If the `int16_t` is < 0 bad things will happen.